src/Functions/FunctionsHashing.h

```diff
@@ -245,12 +246,22 @@ struct HalfMD5Impl
             uint64_t uint64_data;
         } buf;
 
-        MD5_CTX ctx;
-        MD5_Init(&ctx);
-        MD5_Update(&ctx, reinterpret_cast<const unsigned char *>(begin), size);
-        MD5_Final(buf.char_data, &ctx);
+        using EVP_MD_CTX_ptr = std::unique_ptr<EVP_MD_CTX, decltype(&EVP_MD_CTX_free)>;
+        const auto ctx = EVP_MD_CTX_ptr(EVP_MD_CTX_new(), EVP_MD_CTX_free);
```

Comment:
IIUC, `EVP_MD_CTX_new` does a heap allocation (or is this a stack allocation?), and we do it once per hashed value (usually a row). Is it somehow possible to initialize the context just once and reuse it between calls (possibly it can be cleared between invocations, if there is a clear call)?

(same applies to other places in this PR)

Reply:
Let's file a separate improvement issue to handle this.